### DISCUSSION TITLE
Runtime environment variables

### DISCUSSION THREAD

**Discussion by jromero** [buildpacks/community#61] [Ideas]
## Summary

As a developer, I'd like to be able to provide environment variables that would be used at runtime environment when executing a build.

## Details

Currently, `pack` supports providing build environment variables as a method of configuring buildpacks. Runtime environment variables would enable scenarios where it would be prefered for environment variables to be stored within the container image. This would be analogous to the [Dockerfile's `ENV` instruction](https://docs.docker.com/engine/reference/builder/#env) but scoped to only applying at runtime.

Furthermore, it would be ideal if the provided environment variables could have "operations" as [build-time env vars do now](https://github.com/buildpacks/spec/blob/buildpack/0.5/buildpack.md#environment-variable-modification-rules) in certain contexts.

### End-User interactions

Some example extension points to provide this functionality to end-users are as follows:

1. `project.toml`
1. `pack build`
1. `/platform/env.launch/`

#### `project.toml`

```toml
# ...

[[run.env]] # → applicable to all process types
name="<KEY>"
value="<VALUE>"
operation="append|override|..." # optional, default=override

[[run.<process-type>.env]] # → applicable to only <process-type> (most specific wins)
name="<KEY>"
value="<VALUE>"
operation="append|override|..."

# ...
```

#### `pack build`

```
pack build ... --runtime-env <KEY>=<VALUE> --runtime-env-file <path/to/file>
```

NOTES: 
* Pack would simply use the `/platform/env.launch/` contract below to provide configuration to the lifecycle.
* Env vars provided via Pack CLI arguments would **default to `override` operations for all process types**. 
    * For more advances usage, users can provide a `project.toml` file.

#### `/platform/env.launch/`

```
.
└── /platform/env.launch/ → applicable to all process types
    ├── KEY.override
    └── <process-type>/ → applicable to only <process-type> (most specific wins)
        ├── KEY1.override
        ├── KEY2.append
        └── ...
```

#### Configuration

- **At build-time**, the configuration should be read from `<platform>/env.launch/` but stored in some other format (possibly in the `<layers>/env.launch/` directory).
- **At runtime**, the configuration should be read from `<platform>/env.launch/` and applied to the running environment.

#### Order of precedence

This is a proposed order of precedence _(higher takes precedence)_:

1. Container Environment Variables (ie. `docker run --env`, [podspec env vars](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/))
1. [Image Config](https://github.com/opencontainers/image-spec/blob/master/config.md) Environment Variables
    - Ultimately applied as Container Environment Variables
1. Configuration from `<platform>/env.launch/` (provided at runtime)
1. Configuration from `<layers>/env.launch/` (persisted from build-time)
1. Configuration from `<layers>/<layer>/env.launch/` (provided by buildpacks from build-time)

## Alternatives

* Buildpacks can already be used to provide this functionality (see [Paketo's buildpack](https://github.com/paketo-buildpacks/environment-variables))
    * There are multiple downsides to this current approach:
        1. The buildpack must be compatible with a multitude of stacks or each stack author needs to provide a buildpack to enable this functionality.
        2. The interface for adding the env vars aren't as nice (dependant on buildpack implementation) or discoverable.

## Additional Thoughts

1. **BREAKING**: We should align `env` directories with [Buildpack API contract](https://github.com/buildpacks/spec/blob/buildpack/0.5/buildpack.md#build)
    - ~~`<platform>/env` - environment variables for both runtime and build-time (currently this is defined as build-time only)~~
    - `<platform>/env` - **remove** use to prevent unintentionally leaking built-time only env vars
    - `<platform>/env.build` - build-time environment variables
    - `<platform>/env.launch` - runtime environment variables for all process types (this proposal)
    - `<platform>/env.launch/<process-type>/` - runtime environment variables for a specific process type (this proposal)

## References

* Related: https://github.com/buildpacks/pack/issues/914#issuecomment-715517788
* Slack conversation ([join](https://slack.buildpacks.io)): https://buildpacks.slack.com/archives/CCP60GJLS/p1613076546013700

**Comment by jromero** [comment]
Sounds like there's some desire to get this done. This would need to be updated to project descriptor 0.2 and make an RFC.

I've created an issue (https://github.com/buildpacks/rfcs/issues/164) to track this work. Open to contributions. :)

**Comment by sclevine** [comment]
@jromero I agree that runtime env vars in project.toml would be a useful feature.

Why would we expose them in `/platform`? That directory provides read-only configuration to buildpacks, and buildpacks don't need access to runtime env vars.

**Reply by jromero** [reply]
Think of use cases where project.toml isn't in use. For example, a platform such as pack can provide additional configuration `pack build --run-env MY_VAR=adf ...`. The platform in this case needs an interface for providing runtime environment variables similar to how build time environment variables can be provided via `/platform/env`.
